crypto/src/main/scala/org/bitcoins/crypto/Sign.scala

```diff
@@ -32,88 +29,164 @@ trait Sign {
-    * on the input space.
-    *
-    * In short, ALL USES OF THIS FUNCTION THAT SIGN THE SAME DATA WITH DIFFERENT ENTROPY
-    * HAVE THE POTENTIAL TO CAUSE REDUCTIONS IN SECURITY AND PRIVACY, BEWARE!
-    */
-  def signWithEntropyFuture(
+  def asyncSignWithEntropy(
       bytes: ByteVector,
-      entropy: ByteVector): Future[ECDigitalSignature] =
-    signWithEntropyFunction(bytes, entropy)
+      entropy: ByteVector): Future[ECDigitalSignature]
 
-  private def signLowRFuture(bytes: ByteVector, startAt: Long)(implicit
+  private def asyncSignLowR(bytes: ByteVector, startAt: Long)(implicit
       ec: ExecutionContext): Future[ECDigitalSignature] = {
     val sigF: Future[ECDigitalSignature] = if (startAt == 0) {
-      signFunction(bytes)
+      asyncSign(bytes)
     } else {
       val startBytes = ByteVector.fromLong(startAt).padLeft(32).reverse
-      signWithEntropyFunction(bytes, startBytes)
+      asyncSignWithEntropy(bytes, startBytes)
     }
 
     sigF.flatMap { sig =>
       if (sig.bytes.length <= 70) {
         Future.successful(sig)
       } else {
-        signLowRFuture(bytes, startAt + 1)
+        asyncSignLowR(bytes, startAt + 1)
       }
     }
   }
 
-  def signLowRFuture(bytes: ByteVector)(implicit
+  def asyncSignLowR(bytes: ByteVector)(implicit
       ec: ExecutionContext): Future[ECDigitalSignature] = {
-    signLowRFuture(bytes, startAt = 0)
+    asyncSignLowR(bytes, startAt = 0)
+  }
+
+  def publicKey: ECPublicKey
+}
+
+object AsyncSign {
+
+  private case class AsyncSignImpl(
+      asyncSignFunction: ByteVector => Future[ECDigitalSignature],
+      asyncSignWithEntropyFunction: (
+          ByteVector,
+          ByteVector) => Future[ECDigitalSignature],
+      override val publicKey: ECPublicKey)
+      extends AsyncSign {
+
+    override def asyncSign(bytes: ByteVector): Future[ECDigitalSignature] = {
+      asyncSignFunction(bytes)
+    }
+
+    override def asyncSignWithEntropy(
+        bytes: ByteVector,
+        entropy: ByteVector): Future[ECDigitalSignature] = {
+      asyncSignWithEntropyFunction(bytes, entropy)
+    }
   }
 
-  def sign(bytes: ByteVector): ECDigitalSignature = {
-    Await.result(signFuture(bytes), 30.seconds)
+  def apply(
+      asyncSign: ByteVector => Future[ECDigitalSignature],
+      asyncSignWithEntropy: (
+          ByteVector,
+          ByteVector) => Future[ECDigitalSignature],
+      pubKey: ECPublicKey): AsyncSign = {
+    AsyncSignImpl(asyncSign, asyncSignWithEntropy, pubKey)
   }
 
-  def signLowR(bytes: ByteVector)(implicit
-      ec: ExecutionContext): ECDigitalSignature = {
-    Await.result(signLowRFuture(bytes), 30.seconds)
+  def constant(sig: ECDigitalSignature, pubKey: ECPublicKey): AsyncSign = {
+    AsyncSignImpl(_ => Future.successful(sig),
+                  (_, _) => Future.successful(sig),
+                  pubKey)
   }
 
+  /** This dummySign function is useful for the case where we do not have the
```

Comment:
If we go foreword with this refactor, it seems maybe we can eliminate the `signFunction` and just have async signers implement the `AsyncSign` interface?

Reply:
I'm just going to delete this scaladoc as it was copied in and no longer seems to make any sense, I've already eliminated `signFunction` stuff from the actual traits